lib/time/convert.c

```diff
@@ -0,0 +1,334 @@
+}
+
+static void timestamp_add_seconds(int64_t *timestamp, const struct rtc_time *datetime)
+{
+	*timestamp += datetime->tm_sec;
+}
+
+static void timestamp_sub_utc_reference(int64_t *timestamp)
+{
+	*timestamp -= UTC_REFERENCE;
+}
+
+static void timestamp_add_utc_reference(int64_t *timestamp)
+{
+	*timestamp += UTC_REFERENCE;
+}
+
+static void datetime_move_minimum_years(struct rtc_time *datetime, int64_t *timestamp)
+{
+	*timestamp -= TIMESTAMP_MIN;
+	datetime->tm_year += CONFIG_TIME_MIN_YEAR;
+}
+
+static void datetime_move_years(struct rtc_time *datetime, bool *leap_year, int64_t *timestamp)
+{
+	int difference;
+	const struct year_segment *segment;
+
+	*leap_year = false;
+
+	for (uint8_t i = 0; i < ARRAY_SIZE(year_segments); i++) {
+		segment = &year_segments[i];
+
+		if (segment->years == 1) {
+			*leap_year = !segment->repeats;
+		}
+
+		if (*timestamp < segment->seconds) {
+			continue;
+		}
+
+		if (segment->repeats) {
+			difference = (int)(*timestamp / segment->seconds);
+			datetime->tm_year += difference * segment->years;
+			*timestamp %= segment->seconds;
+		} else {
+			datetime->tm_year += segment->years;
+			*timestamp -= segment->seconds;
+		}
+	}
+}
+
+static void datetime_move_months(struct rtc_time *datetime, bool leap_year, int64_t *timestamp)
+{
+	uint32_t days;
+	uint32_t seconds;
+
+	datetime->tm_mon = 0;
+
+	for (uint8_t i = 0; i < ARRAY_SIZE(days_in_month); i++) {
+		days = days_in_month[i];
+
+		if (i == 1 && leap_year) {
+			days++;
+		}
+
+		seconds = days * SECONDS_IN_DAY;
+
+		if (*timestamp < seconds) {
+			return;
+		}
+
+		datetime->tm_mon++;
+		*timestamp -= seconds;
+	}
+}
+
+static void datetime_move_days(struct rtc_time *datetime, int64_t *timestamp)
+{
+	datetime->tm_mday = (int)(1 + (*timestamp / SECONDS_IN_DAY));
+	*timestamp %= SECONDS_IN_DAY;
+}
+
+static void datetime_move_hours(struct rtc_time *datetime, int64_t *timestamp)
+{
+	datetime->tm_hour = (int)(*timestamp / SECONDS_IN_HOUR);
+	*timestamp %= SECONDS_IN_HOUR;
+}
+
+static void datetime_move_minutes(struct rtc_time *datetime, int64_t *timestamp)
+{
+	datetime->tm_min = (int)(*timestamp / SECONDS_IN_MINUTE);
+	*timestamp %= SECONDS_IN_MINUTE;
+}
+
+static void datetime_move_seconds(struct rtc_time *datetime, int64_t *timestamp)
+{
+	datetime->tm_sec = (int)*timestamp;
+	*timestamp = 0;
+}
+
+bool sys_time_validate_timestamp(const int64_t *timestamp)
+{
+	return *timestamp >= UTC_MIN && *timestamp < UTC_MAX;
+}
+
+bool sys_time_validate_datetime(const struct rtc_time *datetime)
+{
+	return datetime->tm_year >= (CONFIG_TIME_MIN_YEAR - 1900) &&
+	       datetime->tm_year <= (CONFIG_TIME_MAX_YEAR - 1900) &&
+	       datetime->tm_mon <= 11 &&
+	       datetime->tm_mday >= 1 &&
+	       datetime->tm_mday <= 31 &&
+	       datetime->tm_hour >= 0 &&
+	       datetime->tm_hour <= 23 &&
+	       datetime->tm_min >= 0 &&
+	       datetime->tm_min <= 59 &&
+	       datetime->tm_sec >= 0 &&
+	       datetime->tm_sec <= 59;
```

Comment:
So, this isn't something called only when e.g. CONFIG_ASSERT=y, it's on the path to every call.  And this is actually a lot of code.  What is it trying to protect agaist?  If the user passes an incorrectly constructed date, it's IMHO perfectly legal to hand them back a nonsense timestamp.  Likewise if the RTC driver gives you a broken struct, it's not your problem to solve here.